packages/components/src/SegmentedControl/SegmentedControlProvider.tsx

```diff
@@ -0,0 +1,60 @@
+import React, {
+  ChangeEventHandler,
+  PropsWithChildren,
+  createContext,
+  useContext,
+  useState,
+} from "react";
+
+interface SegmentedControlProviderContext {
+  handleChange: ChangeEventHandler<HTMLInputElement>;
+  selectedOption: unknown;
+  setSelectedOption: (option: string) => void;
+}
+
+export const SegmentedControlContext =
+  createContext<SegmentedControlProviderContext>({
+    handleChange: () => ({}),
+    selectedOption: null,
+    setSelectedOption: () => ({}),
+  });
+
+interface SegmentedControlProviderProps<T> extends PropsWithChildren {
+  readonly onSelectOption: (value: T) => void;
+  readonly defaultOption: T;
+}
+
+export const useSegmentedControl = () => {
+  const { handleChange, selectedOption, setSelectedOption } = useContext(
+    SegmentedControlContext,
+  );
+
+  return { handleChange, selectedOption, setSelectedOption };
+};
+
+export function SegmentedControlProvider<T>({
+  children,
+  onSelectOption,
+  defaultOption,
+}: SegmentedControlProviderProps<T>) {
+  const [selectedOption, setSelectedOption] = useState<T>(defaultOption);
```

Comment:
Anyway we can change this to be controlled and uncontrolled? Even lifting the concept from headless libraries. https://github.com/radix-ui/primitives/blob/main/packages/react/use-controllable-state/src/useControllableState.tsx